evm_loader/program/src/account/treasury.rs

```diff
@@ -32,3 +36,26 @@ impl<'a> Deref for Treasury<'a> {
     }
 }
 
+impl<'a> MainTreasury<'a> {
+    pub fn from_account(info: &'a AccountInfo<'a>) -> Result<Self, ProgramError> {
+        let expected_key = Pubkey::create_with_seed(
+            &collateral_pool_base::id(),
+            collateral_pool_base::MAIN_BALANCE_SEED,
+            &spl_token::id())?;
+
+        if *info.key != expected_key {
+            return Err!(ProgramError::InvalidArgument; "Account {} - invalid main treasure account", info.key);

```

Comment:
Please also check that owner of this account is spl-token program and this is initialized wrapped SOL token.

Reply:
Fixed